packages/jest-util/src/globsToMatcher.ts

```diff
@@ -44,13 +46,17 @@ export default function globsToMatcher(globs: Array<Config.Glob>): Matcher {
 
   const matchers = globs.map(glob => {
     if (!globsToMatchersMap.has(glob)) {
-      const isMatch = picomatch(glob, picomatchOptions, true);
+      const negated = glob.startsWith('!');
+      const isMatch = picomatch(
+        negated ? glob.slice(1) : glob,
```

Comment:
This `glob.slice(1)` is the fix for the "wrong" negation behavior mentioned in the main PR text.
By removing the exclamation mark we explicitly only match ignored path but later know because of the `negated` flag if a match means ignore this file or not.